core/src/main/java/org/apache/iceberg/PartitionSpecUpdate.java

```diff
@@ -0,0 +1,233 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.transforms.Transforms;
+
+/**
+ * PartitionSpec evolution API implementation.
+ */
+class PartitionSpecUpdate implements UpdatePartitionSpec {
+
+  private final TableMetadata base;
+  private final TableOperations ops;
+  private final List<Consumer<PartitionSpec.Builder>> newSpecFields = new ArrayList<>();
+  private final Map<String, PartitionField> curSpecFields;
+
+  PartitionSpecUpdate(TableOperations ops) {
+    this.ops = ops;
+    this.base = ops.current();
+    this.curSpecFields = base.spec().fields().stream().collect(
+        Collectors.toMap(
+            PartitionField::name,
+            Function.identity(),
+            (n1, n2) -> {
+              throw new IllegalStateException(String.format("Duplicate partition field found: %s", n1));
+            },
+            LinkedHashMap::new
+        )
+    );
+  }
+
+  @Override
+  public PartitionSpec apply() {
+    PartitionSpec.Builder specBuilder = PartitionSpec.builderFor(base.schema());
+    curSpecFields.values().forEach(field ->
+        specBuilder.add(
+            field.sourceId(),
+            field.fieldId(),
+            field.name(),
+            field.transform().toString())
+    );
+    newSpecFields.forEach(c -> c.accept(specBuilder));
+    return specBuilder.build();
+  }
+
+  @Override
+  public UpdatePartitionSpec clear() {
+    newSpecFields.clear();
+    curSpecFields.clear();
```

Comment:
This should call `removeField` for every current field so that v1 placeholders are inserted.